File: react-native-nitro-player/android/src/main/java/com/margelo/nitro/nitroplayer/HybridEqualizer.kt
package com.margelo.nitro.nitroplayer

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.NullType
import com.margelo.nitro.nitroplayer.equalizer.EqualizerCore

@DoNotStrip
@Keep
class HybridEqualizer : HybridEqualizerSpec() {
    private val core: EqualizerCore

    init {
        val context =
            NitroModules.applicationContext ?: throw IllegalStateException("React Context is not initialized")

        // Get the equalizer core - it will initialize lazily with audio session 0
        // and be re-initialized with the proper session when onAudioSessionIdChanged fires
        core = EqualizerCore.getInstance(context)
        core.ensureInitialized()
    }

    @DoNotStrip
    @Keep
    override fun setEnabled(enabled: Boolean): Boolean = core.setEnabled(enabled)

    @DoNotStrip
    @Keep
    override fun isEnabled(): Boolean = core.isEnabled()

    @DoNotStrip
    @Keep
    override fun getBands(): Array<EqualizerBand> = core.getBands()

    @DoNotStrip
    @Keep
    override fun setBandGain(
        bandIndex: Double,
        gainDb: Double,
    ): Boolean = core.setBandGain(bandIndex.toInt(), gainDb)

    @DoNotStrip
    @Keep
    override fun setAllBandGains(gains: DoubleArray): Boolean = core.setAllBandGains(gains)

    @DoNotStrip
    @Keep
    override fun getBandRange(): GainRange = core.getBandRange()

    @DoNotStrip
    @Keep
    override fun getPresets(): Array<EqualizerPreset> = core.getPresets()

    @DoNotStrip
    @Keep
    override fun getBuiltInPresets(): Array<EqualizerPreset> = core.getBuiltInPresets()

    @DoNotStrip
    @Keep
    override fun getCustomPresets(): Array<EqualizerPreset> = core.getCustomPresets()

    @DoNotStrip
    @Keep
    override fun applyPreset(presetName: String): Boolean = core.applyPreset(presetName)

    @DoNotStrip
    @Keep
    override fun getCurrentPresetName(): Variant_NullType_String {
        val name = core.getCurrentPresetName()
        return if (name != null) {
            Variant_NullType_String.create(name)
        } else {
            Variant_NullType_String.create(NullType.NULL)
        }
    }

    @DoNotStrip
    @Keep
    override fun saveCustomPreset(name: String): Boolean = core.saveCustomPreset(name)

    @DoNotStrip
    @Keep
    override fun deleteCustomPreset(name: String): Boolean = core.deleteCustomPreset(name)

    @DoNotStrip
    @Keep
    override fun getState(): EqualizerState = core.getState()

    @DoNotStrip
    @Keep
    override fun reset() = core.reset()

    override fun onEnabledChange(callback: (enabled: Boolean) -> Unit) {
        core.addOnEnabledChangeListener(callback)
    }

    override fun onBandChange(callback: (bands: Array<EqualizerBand>) -> Unit) {
        core.addOnBandChangeListener(callback)
    }

    override fun onPresetChange(callback: (presetName: Variant_NullType_String?) -> Unit) {
        core.addOnPresetChangeListener(callback)
    }
}

File: react-native-nitro-player/ios/HybridEqualizer.swift
//
//  HybridEqualizer.swift
//  NitroPlayer
//
//  Created by Ritesh Shukla on 04/02/26.
//

import Foundation
import NitroModules

final class HybridEqualizer: HybridEqualizerSpec {
  // MARK: - Properties

  private let core: EqualizerCore

  // MARK: - Initialization

  override init() {
    core = EqualizerCore.shared
    super.init()
  }

  // MARK: - Enable/Disable

  func setEnabled(enabled: Bool) throws -> Bool {
    return core.setEnabled(enabled)
  }

  func isEnabled() throws -> Bool {
    return core.isEnabled()
  }

  // MARK: - Band Control

  func getBands() throws -> [EqualizerBand] {
    return core.getBands()
  }

  func setBandGain(bandIndex: Double, gainDb: Double) throws -> Bool {
    return core.setBandGain(bandIndex: Int(bandIndex), gainDb: gainDb)
  }

  func setAllBandGains(gains: [Double]) throws -> Bool {
    return core.setAllBandGains(gains)
  }

  func getBandRange() throws -> GainRange {
    return core.getBandRange()
  }

  // MARK: - Presets

  func getPresets() throws -> [EqualizerPreset] {
    return core.getPresets()
  }

  func getBuiltInPresets() throws -> [EqualizerPreset] {
    return core.getBuiltInPresets()
  }

  func getCustomPresets() throws -> [EqualizerPreset] {
    return core.getCustomPresets()
  }

  func applyPreset(presetName: String) throws -> Bool {
    return core.applyPreset(presetName)
  }

  func getCurrentPresetName() throws -> Variant_NullType_String {
    if let name = core.getCurrentPresetName() {
      return .second(name)
    } else {
      return .first(NullType.null)
    }
  }

  func saveCustomPreset(name: String) throws -> Bool {
    return core.saveCustomPreset(name)
  }

  func deleteCustomPreset(name: String) throws -> Bool {
    return core.deleteCustomPreset(name)
  }

  // MARK: - State

  func getState() throws -> EqualizerState {
    return core.getState()
  }

  func reset() throws {
    core.reset()
  }

  // MARK: - Event Callbacks

  func onEnabledChange(callback: @escaping (Bool) -> Void) throws {
    NitroPlayerLogger.log("HybridEqualizer", "onEnabledChange callback registered")
    core.addOnEnabledChangeListener(owner: self, callback)
  }

  func onBandChange(callback: @escaping ([EqualizerBand]) -> Void) throws {
    NitroPlayerLogger.log("HybridEqualizer", "onBandChange callback registered")
    core.addOnBandChangeListener(owner: self, callback)
  }

  func onPresetChange(callback: @escaping (Variant_NullType_String?) -> Void) throws {
    NitroPlayerLogger.log("HybridEqualizer", "onPresetChange callback registered")
    core.addOnPresetChangeListener(owner: self, callback)
  }
}
